### PR TITLE
Preset: Idiomatic

### PR DESCRIPTION
**Pull request by hzoo**
#1065. WIP

Idiomatic | JSCS
-------------|---------
`indentation` | no preference but be consistent!
`quotes?` | use jquery preset: `"validateQuoteMarks": { "mark": "\"", "escape": true }`,
`foo("bar")`| **new rule option**: `requireSpacesInsideParentheses`: Single argument string literal, no space
`if ( !("foo" in obj) ) {` | **new rule option**: `requireSpacesInsideParentheses`: Inner grouping parens, no space
`foo([ "alpha", "beta" ]);` | `"requireSpacesInsideParentheses": { "except": [ "[", "]" ] }`
`(function( global ) {})()` | `"requireSpacesInsideParentheses": { "except": ["function"] }`
`foo({ a: "alpha" });` | `"requireSpacesInsideParentheses": { "except": [ "{", "}" ] }`
`square( 10 );` | `"disallowSpacesInCallExpression": true`
`// statements` | `"requireSpaceAfterLineComment": { "allExcept": ["#", "="] }`
 | `"disallowMultipleLineBreaks": null`
`var, const, let statements should always be in the beginning of their respective scope` | #99 **new rule** `requireVarDeclFirst`
`name of regex: rDesc` | **new rule**?
`End of line comments are prohibited!` | **new rule** `validateCommentPosition`
`var factorial = function factorial( number ) {` | similar to #850
`avoid switch if possible` | `disallowNodeTypes: ["SwitchStatement"]`
`early returns` | new rule https://github.com/jscs-dev/node-jscs/issues/1223 `requireEarlyReturn`

### CODE REVIEW

**Comment by mdevils** [comment]
+ @markelog

File: presets/idiomatic.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "preset": "jquery",
+    "requireSpacesInsideParentheses": {
+        "all": true,
+        "except": [ "{", "}" ]
```
Comment (markelog) [inline presets/idiomatic.json]: Correct one is 
```js
"except": [ "{", "}", "[", "]", "function" ]
```

And as a said feature, `requireSpacesInsideParentheses` should understand quotes, so case like this:
```js
foo("bar")
```
was acceptable.

File: test/data/options/preset/idiomatic.js [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+// 2.C.1.1
+// Functions with callbacks
+foo(function() {
+  // Note there is no extra space between the first paren
+  // of the executing function call and the word "function"
+});
+
+// Function accepting an array, no space
+foo([ "alpha", "beta" ]);
```
Comment (markelog) [inline test/data/options/preset/idiomatic.js]: If `requireSpacesInsideParentheses` doesn't throw on this, then it's a bug

File: test/data/options/preset/idiomatic.js [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+// 2.C.1.2
+// Function accepting an object, no space
+foo({
```
Comment (markelog) [inline test/data/options/preset/idiomatic.js]: same as above

File: test/data/options/preset/idiomatic.js [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+  a: "alpha",
+  b: "beta"
+});
+
+// Single argument string literal, no space
+foo("bar");
+
+// Inner grouping parens, no space
```
Comment (markelog) [inline test/data/options/preset/idiomatic.js]: Catch cases like this is currently impossible and i think, really hard to implement

File: test/data/options/preset/idiomatic.js [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+if ( !("foo" in obj) ) {
+
+}
+
+// 5.1.1
+// A Practical Module
+
+(function( global ) {
+  var Module = (function() {
```
Comment (markelog) [inline test/data/options/preset/idiomatic.js]: same as above

**Comment by markelog** [comment]
@rwaldron you might like this :-)

File: test/data/options/preset/idiomatic.js (hunk 4 above)
Comment (markelog) [inline test/data/options/preset/idiomatic.js]: Maybe, i overestimated it, maybe not that hard :-), but still need to add additional feature for that rule

File: test/data/options/preset/idiomatic.js (hunk 2 above)
Comment (hzoo) [inline test/data/options/preset/idiomatic.js]: Ok it's good - I was testing on the wrong config or something at the time. Updated the OP

**Comment by hzoo** [comment]
Oh failing because of `requirePaddingNewLinesBeforeLineComments` in jQuery

**Comment by markelog** [comment]
We can't land this, until https://github.com/jscs-dev/node-jscs/pull/1069#discussion_r24805322 is all done, specifically until

```js
foo("bar")
```

 case is handled, would like to take a hit on this one? https://github.com/jscs-dev/node-jscs/issues/836

**Comment by hzoo** [comment]
Yeah I was just updating the PRs since I forgot to update 2 files

**Comment by hzoo** [comment]
Yeah I'l take a look

**Comment by mrjoelkemp** [comment]
What's the status on this PR? Still held up?

**Comment by hzoo** [comment]
Yeah on #836. Need an exception for `foo("bar")` - `requireSpacesInsideParentheses` but except string literals so I guess for
```js
", ', `
```

**Comment by hzoo** [comment]
I guess I can finish this now lol

**Comment by mrjoelkemp** [comment]
@hzoo Can we not just land this with that case in an open issue? Would it be reporting the wrong case or nothing at all? Can we just nullify that case?

Would love to land this finally.

**Comment by hzoo** [comment]
Yeah we could land it now if we want (and make an issue for the other rules that need to be created) @markelog

**Comment by ehartford** [comment]
Please pull?

**Comment by markelog** [comment]
@ehartford 

> still need foo("bar") case

**Comment by ehartford** [comment]
can not make a separate issue for that case?

**Comment by markelog** [comment]
https://github.com/jscs-dev/node-jscs/issues/836

**Comment by hzoo** [comment]
@markelog Do we want to release the new preset now and update it later when #836 is done?

**Comment by markelog** [comment]
Well, we certainly can, but that feature is kind important

File: presets/idiomatic.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{
+    "preset": "jquery",
+    "requireSpacesInsideParentheses": {
+        "all": true,
+        "except": [ "{", "}", "[", "]", "function" ]
+    },
+    "requireSpacesInsideParenthesizedExpression": {
```
Comment (markelog) [inline presets/idiomatic.json]: Why we need this?

**Comment by markelog** [comment]
Can we include `requireVarDeclFirst`?

**Comment by hzoo** [comment]
Ok I can add `requireVarDeclFirst`.

`JSDoc style is good, but requires a significant time investment` - don't think we need to make a rule to prevent jsdoc style unless the rule would be `disallowBlockComments`.

`End of line comments are prohibited!`

That would be https://github.com/jscs-dev/node-jscs/issues/651 - `validateCommentPosition: "above"` (disallow beside)

**Comment by hzoo** [comment]
Getting an error with `requireVarDeclFirst` on 

```js
// Good
function foo() {
  let foo;
  if ( condition ) {
    let bar = "";

  }
}
```

File: presets/idiomatic.json (hunk 6 above)
Comment (hzoo) [inline presets/idiomatic.json]: Thought it was for

```
if ( !("foo" in obj) ) {

}
``` but I guess not

**Comment by markelog** [comment]
> JSDoc style is good, but requires a significant time investment - don't think we need to make a rule to prevent jsdoc style unless the rule would be disallowBlockComments.

I was thinking we should include jsdoc rules 

> Getting an error with requireVarDeclFirst on

https://github.com/jscs-dev/node-jscs/issues/1783

**Comment by hzoo** [comment]
Ok updating with

```js
   "requireSpacesInsideParentheses": {
        "all": true,
        "ignoreParenthesizedExpression": true,
        "except": [ "\"" ]
    },
```

**Comment by markelog** [comment]
How about `disallowSpacesInsideParenthesizedExpression: true` and `validateCommentPosition:`?

**Comment by hzoo** [comment]
Getting

```js
Illegal space after opening grouping parenthesis at 1.js :
   208 |
   209 |  Ctor.prototype.setFoo = function( val ) {
   210 |    return ( this.foo = val );
--------------------^
   211 |  };
   212 |

Illegal space before closing grouping parenthesis at 1.js :
   208 |
   209 |  Ctor.prototype.setFoo = function( val ) {
   210 |    return ( this.foo = val );
-----------------------------------^
   211 |  };
   212 |
```

for disallowSpacesInsideParenthesizedExpression - adding validatecomment

**Comment by markelog** [comment]
:-(

**Comment by markelog** [comment]
It seems we all good here?

**Comment by hzoo** [comment]
Yeah only `requireVarDeclFirst` is left? (and we have an issue for that already)

**Comment by hzoo** [comment]
It was kind of annoying to depend on a preset (jquery) though - since if you update the jquery preset, then the tests will fail (validateIdentation errors in that case) so I just removed that?

**Comment by markelog** [comment]
> since if you update the jquery preset, then the tests will fail (validateIdentation errors in that case) so I just removed that?

Do what you think is right :-).

**Comment by markelog** [comment]
How we doing here?

**Comment by hzoo** [comment]
Can merge - there's only some rules we haven't implemented yet.

**Comment by markelog** [comment]
I guess we did catch the spirit of this preset, we can take care the details afterwards

**Comment by hzoo** [comment]
So :+1: for this?

**Comment by markelog** [comment]
LGTM

**Comment by hzoo** [comment]
Ok finally merged! :tada:

**Pull request closed without merging**